tox.ini

```diff
@@ -7,14 +7,20 @@ toxworkdir={env:TOX_WORK_DIR:.tox}
 
 [testenv]
 deps =
+	pytest
 	# workaround for sphinx-doc/sphinx#9562
 	# TODO: remove after Sphinx>4.1.2 is available.
 	sphinx@git+https://github.com/sphinx-doc/sphinx; python_version>="3.10"
 commands =
-	pytest {posargs}
+	# We invoke pytest like this to avoid the warning:
+	#   UserWarning: Distutils was imported before Setuptools,
+	#   but importing Setuptools also replaces the `distutils`
+	#   module in `sys.modules`.
+	pytest -c 'import setuptools; import pytest; pytest.main("{posargs}".split())'
```

Comment:
I'm uneasy about this change. I'd like to know more about why distutils is imported in pytest and if it's something that can be fixed.

Reply:
+1 if you can find a better solution

Reply:
Here's the error I see:

```
setuptools github_workflows_SETUPTOOLS_USE_DISTUTILS $ $SETUPTOOLS_USE_DISTUTILS='local' ./.tox/python/bin/pytest
Traceback (most recent call last):
  File "/Users/jaraco/code/public/pypa/setuptools/./.tox/python/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 185, in console_main
    code = main()
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 143, in main
    config = _prepareconfig(args, plugins)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 318, in _prepareconfig
    config = pluginmanager.hook.pytest_cmdline_parse(
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_callers.py", line 55, in _multicall
    gen.send(outcome)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/helpconfig.py", line 100, in pytest_cmdline_parse
    config: Config = outcome.get_result()
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1003, in pytest_cmdline_parse
    self.parse(args)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1283, in parse
    self._preparse(args, addopts=addopts)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1191, in _preparse
    self.hook.pytest_load_initial_conftests(
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_hooks.py", line 265, in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_manager.py", line 80, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_callers.py", line 60, in _multicall
    return outcome.get_result()
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_result.py", line 60, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pluggy/_callers.py", line 39, in _multicall
    res = hook_impl.function(*args)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 1070, in pytest_load_initial_conftests
    self.pluginmanager._set_initial_conftests(early_config.known_args_namespace)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 503, in _set_initial_conftests
    self._try_load_conftest(current, namespace.importmode)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 508, in _try_load_conftest
    self._getconftestmodules(anchor, importmode)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 536, in _getconftestmodules
    mod = self._importconftest(conftestpath, importmode)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 587, in _importconftest
    self.consider_conftest(mod)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 663, in consider_conftest
    self.register(conftestmodule, name=conftestmodule.__file__)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 443, in register
    self.consider_module(plugin)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 669, in consider_module
    self._import_plugin_specs(getattr(mod, "pytest_plugins", []))
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 676, in _import_plugin_specs
    self.import_plugin(import_spec)
  File "/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/_pytest/config/__init__.py", line 703, in import_plugin
    __import__(importspec)
  File "/Users/jaraco/code/public/pypa/setuptools/setuptools/__init__.py", line 9, in <module>
    import _distutils_hack.override  # noqa: F401
  File "/Users/jaraco/code/public/pypa/setuptools/_distutils_hack/override.py", line 1, in <module>
    __import__('_distutils_hack').do_override()
  File "/Users/jaraco/code/public/pypa/setuptools/_distutils_hack/__init__.py", line 68, in do_override
    warn_distutils_present()
  File "/Users/jaraco/code/public/pypa/setuptools/_distutils_hack/__init__.py", line 23, in warn_distutils_present
    warnings.warn(
UserWarning: Distutils was imported before Setuptools, but importing Setuptools also replaces the `distutils` module in `sys.modules`. This may lead to undesirable behaviors or errors. To avoid these issues, avoid using distutils directly, ensure that setuptools is installed in the traditional way (e.g. not an editable install), and/or make sure that setuptools is always imported before distutils.
```

Reply:
Yes, it looked like this without the workaround

Reply:
It doesn't appear that simply importing pytest is the culprit:

```
setuptools github_workflows_SETUPTOOLS_USE_DISTUTILS $ cat .tox/python/bin/pytest
#!/Users/jaraco/code/public/pypa/setuptools/.tox/python/bin/python
# -*- coding: utf-8 -*-
import re
import sys
from pytest import console_main
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(console_main())
setuptools github_workflows_SETUPTOOLS_USE_DISTUTILS $ .tox/python/bin/python -c "import pytest, sys; print('distutils' in sys.modules)"
False
```

Reply:
I don't see anything in the pytest code that imports distutils, so I'm guessing there's another pytest plugin that's getting loaded that imports distutils before importing setuptools.

Reply:
Looks like the culprit is virtualenv:

```
>>> import importlib.metadata as md
>>> for ep in md.entry_points(group='pytest11'):
...   ep.load()
...   print({'distutils'} & set(sys.modules))
... 
<module 'pytest_flake8' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_flake8.py'>
set()
<module 'pytest_enabler' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_enabler/__init__.py'>
set()
<module 'pytest_checkdocs' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_checkdocs/__init__.py'>
set()
<module 'pytest_black' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_black.py'>
set()
<module 'pytest_mypy' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_mypy.py'>
set()
<module 'pytest_shutil.workspace' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_shutil/workspace.py'>
set()
<module 'xdist.plugin' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/xdist/plugin.py'>
set()
<module 'xdist.looponfail' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/xdist/looponfail.py'>
set()
<module 'pytest_virtualenv' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_virtualenv.py'>
{'distutils'}
<module 'pytest_forked' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_forked/__init__.py'>
{'distutils'}
<module 'pytest_cov.plugin' from '/Users/jaraco/code/public/pypa/setuptools/.tox/python/lib/python3.10/site-packages/pytest_cov/plugin.py'>
{'distutils'}
```

Reply:
[confirmed](https://github.com/man-group/pytest-plugins/blob/2929621566d6510849878d913da6509635e666e1/pytest-virtualenv/pytest_virtualenv.py#L5-L8)

Reply:
Confirmed that with https://github.com/man-group/pytest-plugins/pull/189, the problem goes away.